tuf/signed/sign.go

```diff
@@ -13,73 +13,76 @@ package signed
 
 import (
 	"crypto/rand"
-	"fmt"
 
 	"github.com/Sirupsen/logrus"
+	"github.com/docker/notary/trustmanager"
 	"github.com/docker/notary/tuf/data"
 	"github.com/docker/notary/tuf/utils"
 )
 
-// Sign takes a data.Signed and a key, calculated and adds the signature
-// to the data.Signed
-func Sign(service CryptoService, s *data.Signed, keys ...data.PublicKey) error {
-	logrus.Debugf("sign called with %d keys", len(keys))
-	signatures := make([]data.Signature, 0, len(s.Signatures)+1)
-	signingKeyIDs := make(map[string]struct{})
-	ids := make([]string, 0, len(keys))
+// Sign takes a data.Signed and replaces data.Signature with at least
+// minPrimarySignature signatures using primaryKeys and any possible
+// signatures using optionalKeys
+func Sign(service CryptoService, s *data.Signed, primaryKeys []data.PublicKey,
+	minSignatures int, optionalKeys []data.PublicKey) error {
+	logrus.Debugf("sign called with %d/%d required + %d optional keys",
+		minSignatures, len(primaryKeys), len(optionalKeys))
 
 	privKeys := make(map[string]data.PrivateKey)
 
 	// Get all the private key objects related to the public keys
-	for _, key := range keys {
+	missingKeyIDs := []string{}
+	for _, key := range primaryKeys {
 		canonicalID, err := utils.CanonicalKeyID(key)
-		ids = append(ids, canonicalID)
 		if err != nil {
-			continue
+			return err
 		}
 		k, _, err := service.GetPrivateKey(canonicalID)
 		if err != nil {
-			continue
+			if _, ok := err.(trustmanager.ErrKeyNotFound); ok {
+				missingKeyIDs = append(missingKeyIDs, canonicalID)
+				continue
+			}
+			return err
 		}
 		privKeys[key.ID()] = k
 	}
 
-	// Check to ensure we have at least one signing key
-	if len(privKeys) == 0 {
-		return ErrNoKeys{KeyIDs: ids}
+	// Check to ensure we have enough signing keys
+	if len(privKeys) < minSignatures {
```

Comment:
Anyway, such a multi-person signing can be easily supported by setting `minSignatures` to 1 instead of the role’s threshold. The API as implemented in this patch will do the right thing when a bigger threshold is somehow set up (and you are right, this likely means failing—but failing is the right thing to do right now), and it does not preclude future multi-signatures.